File: backend/src/auth/auth.service.ts
import { Injectable, Inject, UnauthorizedException, BadRequestException, ForbiddenException, NotFoundException, Logger } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { eq, and } from 'drizzle-orm';
import { PostgresJsDatabase } from 'drizzle-orm/postgres-js';
import * as bcrypt from 'bcrypt';
import * as crypto from 'crypto';
import * as OTPLib from 'otplib';
import * as QRCode from 'qrcode';
import * as schema from '../database/schema';
import { users } from '../database/schema/users';
import { recoveryCodes } from '../database/schema/recovery-codes';
import { DRIZZLE } from '../database/drizzle.provider';
import { AuditService } from '../audit/audit.service';

type Database = PostgresJsDatabase<typeof schema>;

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    @Inject(DRIZZLE) private db: Database,
    private jwtService: JwtService,
    private auditService: AuditService,
  ) {}

  private generateToken(user: { id: string; email: string; role: string }) {
    return this.jwtService.sign({
      sub: user.id,
      email: user.email,
      role: user.role,
    });
  }

  async loginInit(email: string) {
    const [user] = await this.db
      .select({
        id: users.id,
        email: users.email,
        fullName: users.fullName,
        role: users.role,
        isActive: users.isActive,
        isDeleted: users.isDeleted,
        totpSecret: users.totpSecret,
        totpVerified: users.totpVerified,
      })
      .from(users)
      .where(eq(users.email, email))
      .limit(1);

    if (!user) throw new UnauthorizedException('Invalid credentials');
    if (user.isDeleted) throw new UnauthorizedException('Account has been removed');
    if (!user.isActive) throw new UnauthorizedException('Account is deactivated');

    if (user.totpSecret && user.totpVerified) {
      return { enrolled: true };
    }

    const secret = OTPLib.generateSecret();
    const otpUri = OTPLib.generateURI({ issuer: 'KNAPS Docs', label: user.email, secret });
    const qrDataUrl = await QRCode.toDataURL(otpUri);

    await this.db
      .update(users)
      .set({ totpSecret: secret, updatedAt: new Date() })
      .where(eq(users.id, user.id));

    const enrollToken = this.jwtService.sign({
      sub: user.id,
      purpose: 'enroll',
    }, { expiresIn: '15m' });

    return { enrolled: false, enrollToken, qrDataUrl };
  }

  async login(email: string, totpCode: string) {
    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.email, email))
      .limit(1);

    if (!user) throw new UnauthorizedException('Invalid credentials');
    if (user.isDeleted) throw new UnauthorizedException('Account has been removed');
    if (!user.isActive) throw new UnauthorizedException('Account is deactivated');
    if (!user.totpSecret || !user.totpVerified) throw new UnauthorizedException('TOTP not enrolled');

    const result = OTPLib.verifySync({ token: totpCode, secret: user.totpSecret });
    if (!result.valid) throw new UnauthorizedException('Invalid verification code');

    const token = this.generateToken(user);

    this.auditService.log({
      userId: user.id,
      action: 'LOGIN',
      entityType: 'auth',
      entityId: user.id,
      description: `${user.fullName} logged in`,
      newValues: { email: user.email },
      ipAddress: null,
      userAgent: null,
    }).catch((err) => this.logger.error('Failed to log login audit', err));

    return {
      token,
      recoveryCodesMissing: user.role === 'admin' ? !(await this.checkRecoveryCodesStatus(user.id)) : false,
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
        role: user.role,
      },
    };
  }

  async loginWithRecovery(email: string, code: string) {
    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.email, email))
      .limit(1);

    if (!user) throw new UnauthorizedException('Invalid credentials');
    if (!user.isActive) throw new UnauthorizedException('Account is deactivated');
    if (user.isDeleted) throw new UnauthorizedException('Account has been removed');

    const allCodes = await this.db
      .select()
      .from(recoveryCodes)
      .where(and(eq(recoveryCodes.userId, user.id), eq(recoveryCodes.isUsed, false)));

    for (const rc of allCodes) {
      const match = await bcrypt.compare(code.trim().toUpperCase(), rc.codeHash);
      if (match) {
        await this.db
          .update(recoveryCodes)
          .set({ isUsed: true })
          .where(eq(recoveryCodes.id, rc.id));

        const token = this.generateToken(user);
        return {
          token,
          recoveryCodesMissing: user.role === 'admin' ? !(await this.checkRecoveryCodesStatus(user.id)) : false,
          user: { id: user.id, email: user.email, fullName: user.fullName, role: user.role },
        };
      }
    }

    throw new UnauthorizedException('Invalid recovery code');
  }

  async totpEnroll(enrollToken: string, totpCode: string) {
    let payload: { sub: string; purpose: string };
    try {
      payload = this.jwtService.verify(enrollToken);
    } catch {
      throw new UnauthorizedException('Invalid or expired enrollment token');
    }

    if (payload.purpose !== 'enroll') {
      throw new UnauthorizedException('Invalid enrollment token');
    }

    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.id, payload.sub))
      .limit(1);

    if (!user) throw new UnauthorizedException('User not found');
    if (!user.totpSecret) throw new UnauthorizedException('TOTP secret not initialized');

    const resultTotp = OTPLib.verifySync({ token: totpCode, secret: user.totpSecret });
    if (!resultTotp.valid) throw new UnauthorizedException('Invalid verification code');

    await this.db
      .update(users)
      .set({ isActive: true, totpVerified: true, updatedAt: new Date() })
      .where(eq(users.id, user.id));

    const token = this.generateToken(user);

    const recoveryResult = await this.generateRecoveryCodesInternal(user.id);

    this.auditService.log({
      userId: user.id,
      action: 'TOTP_ENROLL',
      entityType: 'auth',
      entityId: user.id,
      description: `${user.fullName} enrolled TOTP authenticator`,
      newValues: { email: user.email },
      ipAddress: null,
      userAgent: null,
    }).catch((err) => this.logger.error('Failed to log TOTP enrollment audit', err));

    return {
      token,
      user: { id: user.id, email: user.email, fullName: user.fullName, role: user.role },
      recoveryCodes: user.role === 'admin' ? recoveryResult : undefined,
    };
  }

  async getEnrollQr(enrollToken: string) {
    let payload: { sub: string; purpose: string };
    try {
      payload = this.jwtService.verify(enrollToken);
    } catch {
      throw new UnauthorizedException('Invalid or expired enrollment token');
    }

    if (payload.purpose !== 'enroll') {
      throw new UnauthorizedException('Invalid enrollment token');
    }

    const [user] = await this.db
      .select({ email: users.email, totpSecret: users.totpSecret })
      .from(users)
      .where(eq(users.id, payload.sub))
      .limit(1);

    if (!user) throw new UnauthorizedException('User not found');
    if (!user.totpSecret) throw new UnauthorizedException('TOTP not initialized');

    const otpUri = OTPLib.generateURI({ issuer: 'KNAPS Docs', label: user.email, secret: user.totpSecret });
    const qrDataUrl = await QRCode.toDataURL(otpUri);

    return { qrDataUrl };
  }

  async reEnroll(userId: string) {
    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);

    if (!user) throw new UnauthorizedException('User not found');

    const secret = OTPLib.generateSecret();
    const otpUri = OTPLib.generateURI({ issuer: 'KNAPS Docs', label: user.email, secret });
    const qrDataUrl = await QRCode.toDataURL(otpUri);

    await this.db
      .update(users)
      .set({ totpSecret: secret, updatedAt: new Date() })
      .where(eq(users.id, userId));

    return { qrDataUrl };
  }

  async adminReEnroll(userId: string, adminId?: string) {
    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);

    if (!user) throw new NotFoundException('User not found');

    await this.db
      .update(users)
      .set({ totpSecret: null, totpVerified: false, updatedAt: new Date() })
      .where(eq(users.id, userId));

    this.auditService.log({
      userId: adminId || null,
      action: 'TOTP_REENROLL',
      entityType: 'user',
      entityId: userId,
      description: `Admin reset authenticator for user '${user.fullName}' (${user.email})`,
      newValues: { email: user.email },
      ipAddress: null,
      userAgent: null,
    }).catch((err) => this.logger.error('Failed to log TOTP admin re-enrollment audit', err));

    return { message: 'Authenticator re-enrollment initiated. User will be prompted to set up on next login.' };
  }

  async reEnrollVerify(userId: string, totpCode: string) {
    const [user] = await this.db
      .select()
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);

    if (!user) throw new UnauthorizedException('User not found');
    if (!user.totpSecret) throw new BadRequestException('No TOTP secret found. Initiate re-enrollment first.');

    const resultTotp = OTPLib.verifySync({ token: totpCode, secret: user.totpSecret });
    if (!resultTotp.valid) throw new BadRequestException('Invalid verification code');

    await this.db
      .update(users)
      .set({ totpVerified: true, updatedAt: new Date() })
      .where(eq(users.id, userId));

    this.auditService.log({
      userId,
      action: 'TOTP_REENROLL',
      entityType: 'auth',
      entityId: userId,
      description: `${user.fullName} re-enrolled TOTP authenticator`,
      newValues: { email: user.email },
      ipAddress: null,
      userAgent: null,
    }).catch((err) => this.logger.error('Failed to log TOTP re-enrollment audit', err));

    return { message: 'Authenticator re-enrolled successfully' };
  }

  private async generateRecoveryCodesInternal(userId: string): Promise<string[]> {
    const newCodes: string[] = [];
    const inserts: { userId: string; codeHash: string }[] = [];

    for (let i = 0; i < 5; i++) {
      const code = crypto.randomBytes(4).toString('hex').toUpperCase();
      newCodes.push(code);
      inserts.push({ userId, codeHash: await bcrypt.hash(code, 10) });
    }

    await this.db.insert(recoveryCodes).values(inserts);
    return newCodes;
  }

  async getRecoveryCodes(userId: string) {
    const codes = await this.db
      .select({ id: recoveryCodes.id, isUsed: recoveryCodes.isUsed, createdAt: recoveryCodes.createdAt })
      .from(recoveryCodes)
      .where(eq(recoveryCodes.userId, userId))
      .orderBy(recoveryCodes.createdAt);

    return codes;
  }

  async generateRecoveryCodes(userId: string) {
    const [targetUser] = await this.db
      .select({ role: users.role })
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);
    if (!targetUser || targetUser.role !== 'admin') {
      throw new ForbiddenException('Recovery codes are only available for admin users');
    }

    await this.db.delete(recoveryCodes).where(eq(recoveryCodes.userId, userId));

    const newCodes: string[] = [];
    const inserts: { userId: string; codeHash: string }[] = [];

    for (let i = 0; i < 5; i++) {
      const code = crypto.randomBytes(4).toString('hex').toUpperCase();
      newCodes.push(code);
      inserts.push({ userId, codeHash: await bcrypt.hash(code, 10) });
    }

    await this.db.insert(recoveryCodes).values(inserts);

    this.auditService.log({
      userId,
      action: 'RECOVERY_CODES',
      entityType: 'auth',
      entityId: userId,
      description: `${(await this.getProfile(userId))?.fullName || 'User'} generated new recovery codes`,
      newValues: { userId },
      ipAddress: null,
      userAgent: null,
    }).catch((err) => this.logger.error('Failed to log recovery codes audit', err));

    return { recoveryCodes: newCodes };
  }

  async checkRecoveryCodesStatus(userId: string): Promise<boolean> {
    const [user] = await this.db
      .select({ role: users.role })
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);

    if (!user || user.role !== 'admin') return true;

    const unused = await this.db
      .select({ id: recoveryCodes.id })
      .from(recoveryCodes)
      .where(and(eq(recoveryCodes.userId, userId), eq(recoveryCodes.isUsed, false)))
      .limit(1);

    return unused.length > 0;
  }

  async getProfile(userId: string) {
    const [user] = await this.db
      .select({
        id: users.id,
        email: users.email,
        fullName: users.fullName,
        role: users.role,
        isActive: users.isActive,
        createdAt: users.createdAt,
      })
      .from(users)
      .where(eq(users.id, userId))
      .limit(1);

    return user;
  }
}

File: backend_python/app/services/auth_service.py
from datetime import timedelta
from uuid import UUID

from sqlmodel import delete, select
from sqlmodel.ext.asyncio.session import AsyncSession

from app.core.config import settings
from app.core.security import (
    blind_index,
    create_access_token,
    generate_qr_code,
    generate_recovery_codes,
    generate_totp_secret,
    get_totp_uri,
    hash_recovery_code,
    verify_recovery_code,
    verify_token,
    verify_totp_code,
)
from app.models.recovery_code import RecoveryCode
from app.models.user import User


class AuthService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def login_init(self, email: str) -> dict:
        email_hash = blind_index(email)
        result = await self.db.exec(
            select(User).where(User.email_hash == email_hash, not User.is_deleted)
        )
        user = result.first()

        if not user:
            return {"enrolled": False}

        if not user.is_active and user.totp_verified:
            return {"enrolled": True, "needs_confirm": True}

        if not user.totp_verified:
            totp_secret = generate_totp_secret()
            totp_uri = get_totp_uri(totp_secret, email)
            qr_data_url = generate_qr_code(totp_uri)

            user.totp_secret = totp_secret
            self.db.add(user)
            await self.db.flush()

            enroll_token = create_access_token(
                subject=str(user.id),
                email=user.email,
                role=user.role,
                expires_delta=timedelta(minutes=10),
            )

            return {
                "enrolled": False,
                "enroll_token": enroll_token,
                "qr_data_url": qr_data_url,
            }

        return {"enrolled": True}

    async def resume_setup(self, email: str, totp_code: str) -> dict:
        email_hash = blind_index(email)
        result = await self.db.exec(
            select(User).where(User.email_hash == email_hash, not User.is_deleted)
        )
        user = result.first()

        if not user:
            raise ValueError("User not found")

        if user.is_active:
            raise ValueError("User is already active")

        if not user.totp_verified or not user.totp_secret:
            raise ValueError("TOTP not enrolled")

        if not verify_totp_code(user.totp_secret, totp_code):
            raise ValueError("Invalid TOTP code")

        # Delete old unused recovery codes
        await self.db.exec(delete(RecoveryCode).where(RecoveryCode.user_id == user.id))

        # Generate fresh recovery codes
        recovery_codes = generate_recovery_codes(5)
        for code in recovery_codes:
            code_hash = hash_recovery_code(code)
            rc = RecoveryCode(user_id=user.id, code_hash=code_hash)
            self.db.add(rc)
        await self.db.flush()

        confirm_token = create_access_token(
            subject=str(user.id),
            email=user.email,
            role=user.role,
            expires_delta=timedelta(minutes=60),
        )

        return {
            "confirm_token": confirm_token,
            "recovery_codes": recovery_codes,
            "user": {
                "id": str(user.id),
                "email": user.email,
                "full_name": user.full_name,
                "role": user.role,
            },
        }

    async def login(self, email: str, totp_code: str) -> dict:
        email_hash = blind_index(email)
        result = await self.db.exec(
            select(User).where(User.email_hash == email_hash, not User.is_deleted)
        )
        user = result.first()

        if not user:
            raise ValueError("User not found")

        if not user.is_active:
            raise ValueError("Account is deactivated")

        if not user.totp_verified or not user.totp_secret:
            raise ValueError("TOTP not enrolled")

        if not verify_totp_code(user.totp_secret, totp_code):
            raise ValueError("Invalid TOTP code")

        token = create_access_token(
            subject=str(user.id),
            email=user.email,
            role=user.role,
            expires_delta=timedelta(minutes=settings.JWT_ACCESS_TOKEN_EXPIRE_MINUTES),
        )

        recovery_codes_result = await self.db.exec(
            select(RecoveryCode).where(
                RecoveryCode.user_id == user.id, not RecoveryCode.is_used
            )
        )
        recovery_codes_missing = len(recovery_codes_result.all()) < 2

        return {
            "token": token,
            "user": {
                "id": str(user.id),
                "email": user.email,
                "full_name": user.full_name,
                "role": user.role,
            },
            "recovery_codes_missing": recovery_codes_missing,
        }

    async def recovery_login(self, email: str, recovery_code: str) -> dict:
        email_hash = blind_index(email)
        result = await self.db.exec(
            select(User).where(User.email_hash == email_hash, not User.is_deleted)
        )
        user = result.first()

        if not user:
            raise ValueError("User not found")

        if not user.is_active:
            raise ValueError("Account is deactivated")

        codes_result = await self.db.exec(
            select(RecoveryCode).where(
                RecoveryCode.user_id == user.id, not RecoveryCode.is_used
            )
        )
        codes = codes_result.all()

        matched = None
        for code in codes:
            if verify_recovery_code(recovery_code, code.code_hash):
                matched = code
                break

        if not matched:
            raise ValueError("Invalid recovery code")

        matched.is_used = True
        self.db.add(matched)
        await self.db.flush()

        token = create_access_token(
            subject=str(user.id),
            email=user.email,
            role=user.role,
            expires_delta=timedelta(minutes=settings.JWT_ACCESS_TOKEN_EXPIRE_MINUTES),
        )

        remaining = await self.db.exec(
            select(RecoveryCode).where(
                RecoveryCode.user_id == user.id, not RecoveryCode.is_used
            )
        )
        recovery_codes_missing = len(remaining.all()) < 2

        return {
            "token": token,
            "user": {
                "id": str(user.id),
                "email": user.email,
                "full_name": user.full_name,
                "role": user.role,
            },
            "recovery_codes_missing": recovery_codes_missing,
        }

    async def totp_enroll(self, enroll_token: str, totp_code: str) -> dict:
        payload = verify_token(enroll_token)
        if not payload:
            raise ValueError("Invalid or expired enrollment token")

        user_id = payload.get("sub")
        if not user_id:
            raise ValueError("Invalid enrollment token")

        result = await self.db.exec(select(User).where(User.id == UUID(user_id)))
        user = result.first()

        if not user:
            raise ValueError("User not found")

        if not user.totp_secret:
            raise ValueError("TOTP not initialized")

        if not verify_totp_code(user.totp_secret, totp_code):
            raise ValueError("Invalid TOTP code")

        recovery_codes = generate_recovery_codes(5)
        for code in recovery_codes:
            code_hash = hash_recovery_code(code)
            rc = RecoveryCode(user_id=user.id, code_hash=code_hash)
            self.db.add(rc)

        user.totp_verified = True
        user.is_active = True
        self.db.add(user)
        await self.db.flush()

        token = create_access_token(
            subject=str(user.id),
            email=user.email,
            role=user.role,
            expires_delta=timedelta(minutes=settings.JWT_ACCESS_TOKEN_EXPIRE_MINUTES),
        )

        return {
            "token": token,
            "user": {
                "id": str(user.id),
                "email": user.email,
                "full_name": user.full_name,
                "role": user.role,
            },
            "recovery_codes": recovery_codes,
        }

    async def get_qr(self, email: str) -> str:
        email_hash = blind_index(email)
        result = await self.db.exec(
            select(User).where(User.email_hash == email_hash, not User.is_deleted)
        )
        user = result.first()

        if not user:
            raise ValueError("User not found")

        if not user.totp_secret:
            totp_secret = generate_totp_secret()
            user.totp_secret = totp_secret
            self.db.add(user)
            await self.db.flush()

        totp_uri = get_totp_uri(user.totp_secret, email)
        return generate_qr_code(totp_uri)

    async def get_recovery_codes(self, user_id: UUID) -> list[str]:
        result = await self.db.exec(
            select(RecoveryCode).where(
                RecoveryCode.user_id == user_id, not RecoveryCode.is_used
            )
        )
        return [code.code_hash for code in result.all()]

    async def generate_new_recovery_codes(self, user_id: UUID) -> list[str]:
        await self.db.exec(delete(RecoveryCode).where(RecoveryCode.user_id == user_id))

        codes = generate_recovery_codes(5)
        for code in codes:
            code_hash = hash_recovery_code(code)
            rc = RecoveryCode(user_id=user_id, code_hash=code_hash)
            self.db.add(rc)

        await self.db.flush()
        return codes
